### PR TITLE
Fix generating file names for edited files, and use wp_unique_filename()

### PR DESCRIPTION
**Pull request by azaozz**
Follow-up to https://github.com/WordPress/gutenberg/pull/23284.

## Description
Fixes few errors and inconsistencies when editing an already-edited image file, and when creating the new attachment post. Also adds some better errors and some cleanup of vars, names, etc. and still has few TBD/TODO.

## How has this been tested?
When testing please monitor what files are created in the uploads dir (`wp-content/uploads/year/month/`).

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: lib/class-wp-rest-image-editor-controller.php [hunk 1]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
+			// CW / CCW
+			$result = $image_editor->rotate( 0 - $params['rotation'] );
 
-		$image_editor = wp_get_image_editor( $image_path );
-		if ( ! $image_editor->load() ) {
-			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+			// TODO: return the "real" error message?
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_rotation_failed', __( 'Unable to rotate this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:129]: What if it doesn't load and there is no error?

File: lib/class-wp-rest-image-editor-controller.php [hunk 2]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
-		$size = $image_editor->get_size();
+		// Calculate the file name.
+		$image_ext  = pathinfo( $image_file, PATHINFO_EXTENSION );
+		$image_name = wp_basename( $image_file, ".{$image_ext}" );
+
+		// Do not append multiple `-edited` to the file name.
+		// The user may be editing a previously edited image.
+		if ( preg_match( '/-edited(-\d+)?$/', $image_name ) ) {
+			// Remove any `-1`, `-2`, etc. `wp_unique_filename()` will add the proper number.
+			$image_name = preg_replace( '/-edited(-\d+)?$/', '-edited', $image_name );
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:156]: Can't we replace immediately? Do we need to match first?

File: lib/class-wp-rest-image-editor-controller.php [hunk 3]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
 		// Save to disk.
-		$saved = $image_editor->save( $filename );
+		$saved = $image_editor->save( $uploads['path'] . "/$filename" );
 
 		if ( is_wp_error( $saved ) ) {
 			return $saved;
 		}
 
-		// Update attachment details.
+		// Create new attachment post.
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:177]: Should we copy the rest of the attachment data?

File: lib/class-wp-rest-image-editor-controller.php [hunk 4]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
+			return new WP_Error( 'rest_unknown_image_file_type', __( 'Unable to edit this image.', 'gutenberg' ), array( 'status' => 500 ) );
 		}
 
-		// Try and load the image itself.
-		$image_path = get_attached_file( $media_id );
-		if ( empty( $image_path ) ) {
-			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
-		}
+		if ( isset( $params['rotation'] ) ) { // TODO: use `rotate` instead?
+			// CW / CCW
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:123]: Might be good to write it in full :)

File: lib/class-wp-rest-image-editor-controller.php [hunk 5]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
 	public function apply_edits( $request ) {
 		require_once ABSPATH . 'wp-admin/includes/image.php';
 
-		$params = $request->get_params();
+		$params        = $request->get_params();
+		$attachment_id = $params['media_id'];
 
-		$media_id = $params['media_id'];
+		// This also confirms the attachment is an image.
+		$image_file = wp_get_original_image_path( $attachment_id );
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:108]: Might be good to leave a comment saying that this is the original uploaded file.

File: lib/class-wp-rest-image-editor-controller.php [hunk 6]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
-		$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
-		$image_editor->crop( $crop_x, $crop_y, $width, $height );
+		$filename = "{$image_name}.{$image_ext}";
 
-		// TODO: Generate filename based on edits.
-		$target_file = 'edited-' . $meta['original_name'];
+		// Create the uploads sub-directory if needed.
+		$uploads = wp_upload_dir();
 
-		$filename = rtrim( dirname( $image_path ), '/' ) . '/' . $target_file;
```
Comment (ellatrix) [inline lib/class-wp-rest-image-editor-controller.php:149]: Previously we were overriding the last edited file, right?

**Review by ellatrix — APPROVED** [review]
Looks good to me! Just left some questions.

File: lib/class-wp-rest-image-editor-controller.php [hunk 7]
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
...
-		$image_path = get_attached_file( $media_id );
-		if ( empty( $image_path ) ) {
-			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
-		}
+		if ( isset( $params['rotation'] ) ) { // TODO: use `rotate` instead?
+			// CW / CCW
+			$result = $image_editor->rotate( 0 - $params['rotation'] );
 
-		$image_editor = wp_get_image_editor( $image_path );
-		if ( ! $image_editor->load() ) {
```
[the remark this answers is not in the document]
Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:129]: No need. `wp_get_image_editor()` instantiates the (proper) editor and loads the file passed as first arg. It always returns a WP_Error on failure. See https://developer.wordpress.org/reference/functions/wp_get_image_editor/.

File: lib/class-wp-rest-image-editor-controller.php (hunk 2 above)
Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:156]: Need it for the `else` to append`-edited`.

File: lib/class-wp-rest-image-editor-controller.php (hunk 6 above)
Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:149]: Yeah, kind of. Was pretty messy. Overwriting the last edited file if it is in the same uploads subdirectory, but not all sub-sizes of it. If the edited file was smaller, and not all sub-sizes were created, some of the old sub-sizes were left "orphaned".

Better workflow and UX would be to keep all edited files, perhaps let the user know they can delete unused files from the media library.

File: lib/class-wp-rest-image-editor-controller.php (hunk 3 above)
Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:177]: Yeah, may make sense. But then we'll get two (or more) attachments with the same description and caption. In general don't think the image alt text should be carried over as alt has to be contextual. Editing an older, already used image would suggest it is being used in different context. Editing an image just after uploading it won't have alt text.

File: lib/class-wp-rest-image-editor-controller.php [hunk 8]
```diff
@@ -101,90 +98,157 @@ public function permission_callback( $request ) {
...
 		}
 
-		$meta = array( 'original_name' => basename( $media_url ) );
+		if ( isset( $params['x'], $params['y'], $params['width'], $params['height'] ) ) {
+			// Check if the crop dimensions are whitin bounds.
+			if (
+				( $params['x'] >= 0 && $params['x'] < 100 ) &&
+				( $params['y'] >= 0 && $params['y'] < 100 ) &&
+				( $params['width'] >= 1 && $params['width'] <= 100 ) &&
+				( $params['height'] >= 1 && $params['height'] <= 100 )
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:128]: We should put these bounds checks in the parameter args definition. We can use `maximum` to set a max value. You may need to use `exclusiveMaximum` for the x and y.

Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:128]: Sounds good. Was just a bit unsure if floats would work, but with `width` and `height` expecting minimum `1`, `x` and `y` can be set to expect maximum `99` (inclusive). (Cannot have an image cropped to 1% width but starting point be over 99% of the width, etc.)

Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:128]: Thinking a bit more, the API should only care if these numbers are percents, 0 - 100. The rest should be handled when doing the cropping.

File: lib/class-wp-rest-image-editor-controller.php [hunk 9]
```diff
@@ -101,90 +104,149 @@ public function permission_callback( $request ) {
...
 
-		$meta = array( 'original_name' => basename( $media_url ) );
+		if ( isset( $params['x'], $params['y'], $params['width'], $params['height'] ) ) {
+			$crop = true;
+		}
 
-		if ( isset( $attachment_info['richimage'] ) ) {
-			$meta = array_merge( $meta, $attachment_info['richimage'] );
+		if ( ! $rotate && ! $crop ) {
+			$error = __( 'The image was not erdited. Edit the image before applying the changes.', 'gutenberg' );
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:133]:
```suggestion
			$error = __( 'The image was not edited. Edit the image before applying the changes.', 'gutenberg' );
```

Comment (azaozz) [reply lib/class-wp-rest-image-editor-controller.php:133]: Ouch, thanks! Would a 400 be appropriate error there? At this point it's an "empty" request, nothing to do.

Comment (TimothyBJacobs) [reply lib/class-wp-rest-image-editor-controller.php]: Yeah I think a `400` makes sense.

**Review by TimothyBJacobs — APPROVED** [review]

**Pull request merged**